### PR TITLE
raftstore: fix checking for snapshot last index

### PR DESCRIPTION
**Pull request by LintianShi**
Signed-off-by: LintianShi <lintian.shi@pingcap.com>

### What is changed and how it works?
Issue Number: Close https://github.com/tikv/tikv/issues/12618

What's Changed:

```commit-message
using commit instead of last_index to check gap between existing raft logs and snapshot when recovering from applying state.
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- None

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by BusyJay** [comment]
Any test case?

**Comment by LintianShi** [comment]
I've tested `test_snapshot_recover_from_raft_write_failure` in test_snap.rs. I'm trying to create a special test case for this condition.

File: tests/failpoints/cases/test_snap.rs [hunk 1]
```diff
@@ -785,3 +785,109 @@ fn test_snapshot_recover_from_raft_write_failure() {
...
+            .msg_type(MessageType::MsgAppend)
+            .direction(Direction::Recv),
+    ));
+    cluster.add_send_filter(CloneFilterFactory(
+        RegionPacketFilter::new(1, 3)
+            .msg_type(MessageType::MsgAppend)
+            .direction(Direction::Recv),
+    ));
+
+    // Reer 1 appends entries which is never committed.
```
Comment (5kbpers) [inline tests/failpoints/cases/test_snap.rs:824]:
```suggestion
    // Peer 1 appends entries which is never committed.
```

**Review by 5kbpers — APPROVED** [review]
rest LGTM

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore/src/store/peer_storage.rs [hunk 2]
```diff
@@ -464,13 +464,13 @@ pub fn recover_from_applying_state<EK: KvEngine, ER: RaftEngine>(
...
-    // if we recv append log when applying snapshot, last_index in raft_local_state will
-    // larger than snapshot_index. since raft_local_state is written to raft engine, and
-    // raft write_batch is written after kv write_batch, raft_local_state may wrong if
+    // since raft_local_state is written to raft engine, and
+    // raft write_batch is written after kv write_batch. raft_local_state may wrong if
     // restart happen between the two write. so we copy raft_local_state to kv engine
-    // (snapshot_raft_state), and set snapshot_raft_state.last_index = snapshot_index.
-    // after restart, we need check last_index.
-    if last_index(&snapshot_raft_state) > last_index(&raft_state) {
+    // (snapshot_raft_state), and set snapshot_raft_state.hard_state.commit = snapshot_index.
```
Comment (BusyJay) [inline components/raftstore/src/store/peer_storage.rs:470]: The comment is wrong, we don't set `hard_state.commit` explicitly.

File: tests/failpoints/cases/test_snap.rs [hunk 3]
```diff
@@ -785,3 +785,109 @@ fn test_snapshot_recover_from_raft_write_failure() {
...
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    cluster.must_put(b"k1", b"v1");
+    for i in 1..4 {
+        must_get_equal(&cluster.get_engine(i), b"k1", b"v1");
+    }
+
+    // Guarantee that peer 2 and 3 won't receive any entries,
+    // so these entries cannot be committed.
+    cluster.add_send_filter(CloneFilterFactory(
```
Comment (BusyJay) [inline tests/failpoints/cases/test_snap.rs:813]: Why not just
```
RegionPacketFilter::new(1, 1)
            .msg_type(MessageType::MsgAppend)
            .direction(Direction::Send)
```

File: components/raftstore/src/store/peer_storage.rs (hunk 2 above)
Comment (LintianShi) [reply components/raftstore/src/store/peer_storage.rs:470]: In `save_snapshot_raft_state_to`, we set commit of hard_state in snapshot_raft_state to snapshot_index. And then we persist snapshot_raft_state into kvdb.
```
let mut snapshot_raft_state = self.raft_state.clone();
        snapshot_raft_state
            .mut_hard_state()
            .set_commit(snapshot_index);
```

Comment (BusyJay) [reply components/raftstore/src/store/peer_storage.rs:470]: Got it.

**Comment by BusyJay** [comment]
The test case can pass without the fix on my local machine.

**Comment by LintianShi** [comment]
In my dev machine, it panicked at line 878:` cluster.run_node(1).unwrap();`, when peer 1 recovers from crash.
```
running 1 test
test cases::test_snap::test_snapshot_recover_from_raft_write_failure_with_uncommitted_log ... FAILED

failures:

---- cases::test_snap::test_snapshot_recover_from_raft_write_failure_with_uncommitted_log stdout ----
thread 'cases::test_snap::test_snapshot_recover_from_raft_write_failure_with_uncommitted_log' panicked at 'called `Result::unwrap()` 
on an `Err` value: RaftServer(Other("[components/raftstore/src/store/peer_storage.rs:774]: [region 1] 1 validate state fail: 
Other(\"[components/raftstore/src/store/peer_storage.rs:612]: applied index > max(commit index, recorded commit index), 
region 1, raft state hard_state { term: 6 vote: 1 commit: 7 } last_index: 857, apply state applied_index: 33 commit_index: 7 
commit_term: 6 truncated_state { index: 33 term: 8 }\")"))', tests/failpoints/cases/test_snap.rs:878:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: tests/failpoints/cases/test_snap.rs [hunk 4]
```diff
@@ -785,3 +785,109 @@ fn test_snapshot_recover_from_raft_write_failure() {
         cluster.must_put(format!("k1{}", i).as_bytes(), b"v1");
     }
 }
+
+#[test]
+fn test_snapshot_recover_from_raft_write_failure_with_uncommitted_log() {
```
Comment (BusyJay) [inline tests/failpoints/cases/test_snap.rs:790]: Please add comment to explain the case.

File: tests/failpoints/cases/test_snap.rs [hunk 5]
```diff
@@ -785,3 +785,106 @@ fn test_snapshot_recover_from_raft_write_failure() {
         cluster.must_put(format!("k1{}", i).as_bytes(), b"v1");
     }
 }
+
+/// Redo snapshot apply after restart when kvdb state is updated but raftdb state is not.
+/// In this test case, the peer has an extra long uncommitted log when applying snasphot.
```
Comment (BusyJay) [inline tests/failpoints/cases/test_snap.rs:790]:
```suggestion
/// Test whether applying snapshot is resumed properly when last_index before applying snapshot
/// is larger than the snapshot index and applying is aborted between kv write and raft write.
```

**Review by BusyJay — APPROVED** [review]

**Pull request merged**